File: engine/wdtp.py
"""
HashHelix Stage 11 — Canonical WDTP Reference Engine (Python)

Engine-only. No relic / business logic.
LAW 4 (NER) binding:
    phase = (a_{n-1} + pi/n) mod 2pi
"""

from __future__ import annotations
import math
from typing import Iterator, List, Optional


TAU = math.tau  # 2*pi


def ner_phase(prev_a: int, n: int) -> float:
    """
    Numerical Evaluation Rule (LAW 4):
    Reduce phase mod 2π before sin() to prevent drift.
    """
    # phase_raw = prev_a + pi/n
    phase_raw = prev_a + (math.pi / n)
    # Reduce mod 2π deterministically
    return math.fmod(phase_raw, TAU)


def wdtp_step(prev_a: int, n: int) -> int:
    """
    Single WDTP step.
    a_n = floor(n * sin(phase)) + 1
    with NER-compliant phase reduction.
    """
    if n < 2:
        raise ValueError("n must be >= 2")
    phase = ner_phase(prev_a, n)
    return math.floor(n * math.sin(phase)) + 1


def wdtp_sequence(n_max: int, a1: int = 1) -> List[int]:
    """
    Generate WDTP sequence [a1, a2, ..., a_nmax].
    """
    if n_max < 1:
        raise ValueError("n_max must be >= 1")
    seq = [a1]
    prev = a1
    for n in range(2, n_max + 1):
        prev = wdtp_step(prev, n)
        seq.append(prev)
    return seq


def wdtp_iter(start_n: int = 2, a1: int = 1) -> Iterator[int]:
    """
    Infinite generator of WDTP values starting at a1, yielding a1, a2, ...
    """
    if start_n < 2:
        raise ValueError("start_n must be >= 2")
    prev = a1
    yield prev  # a1
    n = start_n
    while True:
        prev = wdtp_step(prev, n)
        yield prev
        n += 1


def wdtp_prefix_hash(n_max: int, a1: int = 1) -> str:
    """
    Convenience for equivalence tests:
    Returns a stable hex digest of the first n_max terms.
    (SHA-256 of comma-joined ints)
    """
    import hashlib
    seq = wdtp_sequence(n_max, a1=a1)
    blob = ",".join(map(str, seq)).encode("utf-8")
    return hashlib.sha256(blob).hexdigest()


if __name__ == "__main__":
    # quick smoke
    print(wdtp_sequence(20))
    print("hash20:", wdtp_prefix_hash(20))

File: engine/rust/wdtp.rs
//! HashHelix Stage 11 — WDTP Rust Skeleton (NER-Compliant)
//! Engine-only. No business / relic logic.
//! LAW 4 (NER) binding:
//!   phase = (a_{n-1} + PI/n) mod 2PI

use std::f64::consts::PI;

pub const TAU: f64 = std::f64::consts::TAU; // 2π

/// Numerical Evaluation Rule (LAW 4):
/// Reduce phase mod 2π before sin() to prevent drift.
pub fn ner_phase(prev_a: i64, n: u64) -> f64 {
    if n < 2 {
        panic!("n must be >= 2");
    }
    let phase_raw = (prev_a as f64) + (PI / (n as f64));
    // Deterministic reduction mod 2π
    phase_raw.rem_euclid(TAU)
}

/// Single WDTP step.
/// a_n = floor(n * sin(phase)) + 1
pub fn wdtp_step(prev_a: i64, n: u64) -> i64 {
    if n < 2 {
        panic!("n must be >= 2");
    }
    let phase = ner_phase(prev_a, n);
    ((n as f64) * phase.sin()).floor() as i64 + 1
}

/// Generate WDTP sequence [a1, a2, ..., a_nmax].
pub fn wdtp_sequence(n_max: u64, a1: i64) -> Vec<i64> {
    if n_max < 1 {
        panic!("n_max must be >= 1");
    }
    let mut seq: Vec<i64> = Vec::with_capacity(n_max as usize);
    seq.push(a1);
    let mut prev = a1;
    for n in 2..=n_max {
        prev = wdtp_step(prev, n);
        seq.push(prev);
    }
    seq
}

/// Iterator over WDTP values (infinite).
pub struct WdtpIter {
    n: u64,
    prev: i64,
}

impl WdtpIter {
    pub fn new(a1: i64) -> Self {
        Self { n: 2, prev: a1 }
    }
}

impl Iterator for WdtpIter {
    type Item = i64;

    fn next(&mut self) -> Option<Self::Item> {
        if self.n == 2 {
            // first yield is a1
            self.n += 1;
            return Some(self.prev);
        }
        self.prev = wdtp_step(self.prev, self.n);
        self.n += 1;
        Some(self.prev)
    }
}

/*
Optional equivalence helper (add sha2 crate later):

pub fn wdtp_prefix_hash(n_max: u64, a1: i64) -> String {
    use sha2::{Digest, Sha256};
    let seq = wdtp_sequence(n_max, a1);
    let blob = seq.iter()
        .map(|v| v.to_string())
        .collect::<Vec<_>>()
        .join(",");
    let mut hasher = Sha256::new();
    hasher.update(blob.as_bytes());
    format!("{:x}", hasher.finalize())
}
*/

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn small_n_vector_matches_python() {
        let seq20 = wdtp_sequence(20, 1);
        let expected: Vec<i64> = vec![
            1, 2, 1, 4, -4, 2, 5, -6, 6, 1,
            11, -11, 13, 9, 4, -13, -4, 12, -7, -10
        ];
        assert_eq!(seq20, expected);
    }
}
